### PR TITLE
Move check for NRVO from ReturnStatement semantic into FuncDeclaration

### PR DESCRIPTION
**Pull request by ibuclaw**
Due to patches such as #7990, there's a lot of logic for determining NRVO by analysing front-end ASTs in the back-end.

GDC only checks whether `nrvo_can` is set, and then only overrides it if the target does not return the struct/static array on the stack.  So there may be a number of missed cases where NRVO is never done by GDC because the DMD front-end sets `nrvo_can = false` - only for the DMD back-end to ignore that later and return it by RVO in the codegen pass.

This moves the logic for setting `nrvo_can` into a function, then mirrors the logic from the DMD back-end to the front-end.

### CODE REVIEW

**Comment by WalterBright** [comment]
This looks ready to me. Why is it labeled as WIP?

**Comment by wilzbach** [comment]
CC @RazvanN7 who is trying to move out logic from the AST nodes.

**Comment by RazvanN7** [comment]
@ibuclaw Is this still something of interest to you? If not, can we close this?

**Comment by thewilsonator** [comment]
Rebased, lets see what CI thinks.

File: src/dmd/func.d [hunk 1]
```diff
@@ -2496,6 +2496,55 @@ extern (C++) class FuncDeclaration : Declaration
             error("parameters must be `main()` or `main(string[] args)`");
     }
 
+    /***********************************************
+     * Check all return statements for a function to verify that returning
+     * using NRVO is possible.
+     *
+     * Returns:
+     *      true if the result cannot be returned by hidden reference.
```
Comment (MoonlightSentinel) [inline src/dmd/func.d]: Returning `true` if NRVO is **not** possible feels kinda wrong

Comment (thewilsonator) [reply src/dmd/func.d]: A lot of the `checkFoo` functions do that. I don't think thats too out of place.

Comment (MoonlightSentinel) [reply src/dmd/func.d]: Fair enough

**Pull request merged**